File: algorithms/[938]二叉搜索树的范围和/solution.ts
// topics = ["树"]

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * Recursion
 * time O(n), space O(n), n 为二叉搜索树的节点数
 */
function rangeSumBST(root: TreeNode | null, low: number, high: number): number {
  let res = 0;

  const helper = (node: TreeNode | null): void => {
    if (!node) return;
    helper(node.left);
    if (node.val > high) {
      return;
    }
    if (low <= node.val && node.val <= high) {
      res += node.val;
    }
    helper(node.right);
  };

  helper(root);
  return res;
}

File: algorithms/[938]二叉搜索树的范围和/solution.py
# topics = ["树"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from __future__ import annotations
from typing import Optional, List, Union


class TreeNode:
    def __init__(
        self,
        x: int = 0,
        left: Optional[TreeNode] = None,
        right: Optional[TreeNode] = None,
    ):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        """
        近似题 94.二叉树的中序遍历
        time O(n), space O(n), n 为二叉搜索树的节点数
        """
        s: List[TreeNode] = []
        it: Union[TreeNode, None] = root
        res = 0

        while it or s:
            while it:
                s.append(it)
                it = it.left

            it = s.pop()
            if it.val > high:
                return res
            if low <= it.val <= high:
                res += it.val
            it = it.right

        return res
